feat: add XML formatting advice and integration tests

hamo225 opened the pull request:
- implement XmlFormattingResponseBodyAdvice to format all API XML responses
- add unit and integration tests
- update norm controller test expectations to handle formatted/self-closing XML

RISDEV-8144

hamo225 committed:
feat: add XML formatting advice and integration tests

- implement XmlFormattingResponseBodyAdvice to format all API XML responses
- add unit and integration tests
- update controller test expectations to handle formatted/self-closing XML

RISDEV-8144

hamo225 committed:
test: fix e2e test isolation in references

RISDEV-8144

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * ResponseBodyAdvice that formats XML responses
+ */
+@ControllerAdvice
```
I would mark the whole new class with _@NullMarked_, as we wanted to do for new classes, following the nullability approach that Sebastian presented a while ago. This way the whole class will be considered not nullable in regards of the fields and input parameters of the methods.
This is a good change to use it BECAUSE the package of spring where the interface ResponseBodyAdvance resides is marked by Spring as `@NonNullApi`, meaning that effectively all parameters of both methods are implicitly non nullable by contract
You must have seen some soner warnings locally at least complaining about all non-nullable parameters with something like _Not annotated parameter overrides @NonNullApi parameter_

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  public Object beforeBodyWrite(
+    Object body,
```
This is the only parameter marked in the interface as @Nullable, so mark it also here. From here:
`import org.springframework.lang.Nullable;`

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+    MethodParameter returnType,
+    Class<? extends HttpMessageConverter<?>> converterType
+  ) {
+    return true;
```
We could already restrict here what types should be processed by this advice, because otherwise it will apply to all responses, including non-XML (though the beforeBodyWrite later is anyway restricting already by media type). And the type we need would be only String, so something like:
`return returnType.getParameterType() == String.class;`

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+      serializer.write(document, lsOutput);
+      return writer.toString();
+    } catch (Exception e) {
+      return xml;
```
Please add logging in case of an exception, otherwise will always fail silently. For that add at the class level the annotation `@Slf4j` and then add a line with `log.error(`) and some descriptive message. Thanks!

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+    ServerHttpResponse response
+  ) {
+    if (
+      selectedContentType != null &&
```
You don't need the checking if selectedContentType is null, once you have marked the class as @NullMarked

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+
+  private String formatXml(String xml) {
+    try {
+      DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
```
To get the `Document` out of the string you could just re-use our `XmlMapper` class, method `toDocument(string)`

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+      LSSerializer serializer = impl.createLSSerializer();
+
+      serializer.getDomConfig().setParameter("format-pretty-print", Boolean.TRUE);
+      serializer.getDomConfig().setParameter("xml-declaration", Boolean.FALSE);
```
Interesting line.... and I just realized that the migrated data does not include the XML declaration but our seeds do. With this line all XMLs returned through our API won't include the XML declaration and I am right now not sure what do we want...

hamo225 replied:
@VictorDelCampo  Should they be the same? If there’s no strong requirement, maybe keeping it matching the migrated data (no declaration) makes things more consistent

VictorDelCampo replied:
I would just keep that xml-declaration parameter set to FALSE

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdvice.java:
```diff
@@ -0,0 +1,80 @@
+    return true;
+  }
+
+  @Override
```
You would also need to mark this method as @Nullable

VictorDelCampo commented on backend/src/test/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/controller/NormExpressionControllerTest.java:
```diff
@@ -126,7 +126,7 @@ class getNormXml {
       // Given
       final String eli = "eli/bund/bgbl-1/1990/s2954/2022-12-19/1/deu/regelungstext-verkuendung-1";
-      final String xml = "<target></target>";
+      final String xml = "</target>";
```
Why this changed was introduced? Is not wrong but I am wondering

hamo225 replied:
As far as I can remember, there was a test rule complaining that empty XML elements shouldn’t use separate open/close tags, so I just switched it to a self-closing form to make it pass

VictorDelCampo commented on backend/src/test/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/formatter/XmlFormattingResponseBodyAdviceTest.java:
```diff
@@ -0,0 +1,62 @@
+    String compactXml =
+      "<?xml version=\"1.0\" encoding=\"UTF-8\"?><akn:doc><akn:meta><akn:identification><akn:FRBRWork><akn:FRBRthis value=\"test\"/></akn:FRBRWork></akn:identification></akn:meta><akn:body><akn:article eId=\"art-1\"><akn:heading>Test</akn:heading></akn:article></akn:body></akn:doc>";
+
+    var result = xmlFormatter.beforeBodyWrite(
```
Not sure if the unit test now will work without problems if we introduced the NullMarked at the class level.. if so we would need to create dummy mock object slike:

import static org.mockito.Mockito.mock;

```
Class<? extends HttpMessageConverter<?>> dummyConverterType = HttpMessageConverter.class;
ServerHttpRequest dummyRequest = mock(ServerHttpRequest.class);
ServerHttpResponse dummyResponse = mock(ServerHttpResponse.class);
```

VictorDelCampo commented on frontend/e2e/application/references.spec.ts:
```diff
@@ -23,18 +23,18 @@ test("should be able to add a new ref and edit it's refersTo and href and delete
     .getByRole("textbox")
 
-  const textElement = container.getByText(/Zur Unterstützung der/)
+  const textElement = container.getByText(/Nach Ablauf der/)
```
I am confused, why did you need to modify this E2E tests?

VictorDelCampo replied:
@hamo225 i still do not understand why you needed to adapt this E2E test

VictorDelCampo commented on backend/src/test/java/de/bund/digitalservice/ris/norms/integration/adapter/input/restapi/XmlFormattingResponseBodyAdviceIntegrationTest.java:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  @Test
+  void itReturnsPrettyPrintedXmlEvenIfStoredUgly() throws Exception {
```
Why are you loading and saving the whole norm with eli/bund/bgbl-1/1964/s593/1964-08-05/1/deu/1964-08-05 ? This already includes the regelungstext under eli/bund/bgbl-1/1964/s593/1964-08-05/1/deu/1964-08-05/regelungstext-verkuendung-1.xml
The endpoint you use in the test /api/v1/norms/{eli}. accepts a dokumentExpressionEli, meaning a `eli/bund/bgbl-1/1964/s593/1964-08-05/1/deu/1964-08-05/regelungstext-verkuendung-1` (without the .xml at the end)
So what you would need is loading the regelungstext, transforming it into a DokumentDto and then ugly its xml before saving it into the db. Then you make the API call and test the response against the before-uglyed XML

```
    Regelungstext prettyXml = Fixtures.loadRegelungstextFromDisk(
      "eli/bund/bgbl-1/1964/s593/1964-08-05/1/deu/1964-08-05/regelungstext-verkuendung-1.xml"
    );
    DokumentDto dokumentDto = DokumentMapper.mapToDto(prettyXml);
    dokumentDto.setXml(dokumentDto.getXml().replaceAll(">\\s+<", "><").replaceAll("\\r?\\n", ""));
    dokumentRepository.save(dokumentDto);
```

Then of course at the end checking, ignoring whitespaces, both XMLs, the API response and the actual XML before you modify it

VictorDelCampo reviewed (dismissed):
Nice initial work in the backend Tarek! I made a thorough review and wrote many comments with lots of information. Read them through and let me know if you wanna a 1on1 sync on the topics :)

hamo225 committed:
chore(xml-formatting): apply PR review feedback

- Update XmlFormattingResponseBodyAdviceTest
- Refactor XmlFormattingIntegrationTest
- Update XmlFormattingResponseBodyAdvice

RISDEV-8144

hamo225 committed:
refactor: simplify XmlFormattingIntegrationTest

RISDEV-8144

VictorDelCampo commented on backend/src/test/java/de/bund/digitalservice/ris/norms/integration/adapter/input/restapi/XmlFormattingResponseBodyAdviceIntegrationTest.java:
```diff
@@ -0,0 +1,80 @@
+    DokumentDto dto = DokumentMapper.mapToDto(originalText);
+    dokumentRepository.save(dto);
+
+    String formattedXml = mockMvc
```
You don't need to call an endpoint to get the XML of the norm. the `DokumentDto` itself has the field `xml`, containing the xml in a string form. This is how I proposed it to you before:

`dokumentDto.setXml(dokumentDto.getXml().replaceAll(">\\s+<", "><").replaceAll("\\r?\\n", ""));`

hamo225 committed:
refactor: simplify XmlFormattingIntegrationTest

- use an xml without comments

RISDEV-8144

VictorDelCampo committed:
Remove xml-declaration paramete (default true)

RISDEV-8144

VictorDelCampo committed:
Merge branch 'main' into feature/pretty-print-xml-response
